sycl/include/CL/__spirv/spirv_vars.hpp

```diff
@@ -47,13 +47,17 @@ __DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationId_x();
 __DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationId_y();
 __DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationId_z();
 
+__DPCPP_SYCL_EXTERNAL size_t __spirv_GlobalLinearId();
+__DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationIndex();
+
```

Comment:
@intel/llvm-reviewers-cuda, I haven't implemented these functions for NVIDIA and AMDGPU targets.
What is the best way to do that?
I think I implement `__spirv_LocalInvocationIndex` myself. It can be done by adding implementation to libclc. 
`__spirv_GlobalLinearId` concerns me because it uses offset to compute linear id. I know that offset in SYCL is resolved in the compiler pass, but I'm not sure how offset is exposed to the OpenCL C language. I would appreciate your help with `__spirv_GlobalLinearId` implementation.

Reply:
> I think that __spirv_GlobalLinearId maps to e.g. `__nvvm_read_ptx_sreg_tid_x` from
> 
> https://github.com/intel/llvm/blob/dc0848d35984d1d1d27739d8b73c8c275387bec9/clang/include/clang/Basic/BuiltinsNVPTX.def#L88
> 
> 
> which corresponds to the ptx:
> https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#special-registers-tid
> 
> I don't know about offset issues.

`__nvvm_read_ptx_sreg_tid_x` corresponds to `LocalIvocationId` built-in variable in SPIR-V (or `get_local_id` in OpenCL).

Let me switch to OpenCL terminology, which might be more convenient.
What I'm looking for is to express `get_local_linear_id` and `get_global_linear_id`.
I know how to compute `get_local_linear_id`. It's just get_local_id(2) * get_local_size(1) * get_local_size(0) + get_local_id(1) * get_local_size(0) + get_local_id(0); or using CUDA terms threadIdx.x + threadIdx.y * blockDim.x + threadIdx.z * blockDim.y * blockDim.x. Should be easy to implement libclc.

The problem is `get_global_linear_id`. For 1-dim case the value is get_global_id(0) - get_global_offset(0). There is "offset" concept in CUDA, so we added a new built-in recognized by SYCL compiler (`llvm.{amdgcn|nvvm}.implicit.offset` in llvm), which is resolved at compile time by [GlobalOffset](https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/SYCLLowerIR/GlobalOffset.h) pass. The header file for the pass says that the intrinsic is exposed to SYCL, but libclc is written in OpenCL.

Do I overcomplicate things and can use standard OpenCL built-ins in libclc to implement SYCL functions?

@Naghasan, FYI.

Reply:
you are a bit @bader , the intrinsics is exposed indirectly and offset is handled in libclc for nvptx and amd.

I think someone pointed out already you can't use `__spirv_GlobalLinearId` for SYCL as I can't find the code in the current version.